File: android/src/main/kotlin/com/snowplowanalytics/snowplow_tracker/readers/configurations/GdprConfigurationReader.kt
// Copyright (c) 2022 Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

package com.snowplowanalytics.snowplow_tracker.readers.configurations

import com.snowplowanalytics.snowplow.configuration.GdprConfiguration
import com.snowplowanalytics.snowplow.util.Basis

class GdprConfigurationReader(val values: Map<String, Any>) {
    val basisForProcessing: String by values
    val documentId: String by values
    val documentVersion: String by values
    val documentDescription: String by values

    fun toConfiguration(): GdprConfiguration {
        val basis = when (basisForProcessing) {
            "contract" -> Basis.CONTRACT
            "legal_obligation" -> Basis.LEGAL_OBLIGATION
            "legitimate_interests" -> Basis.LEGITIMATE_INTERESTS
            "public_task" -> Basis.PUBLIC_TASK
            "vital_interests" -> Basis.VITAL_INTERESTS
            else -> Basis.CONSENT
        }
        return GdprConfiguration(basis, documentId, documentVersion, documentDescription)
    }
}

File: ios/snowplow_tracker/Sources/snowplow_tracker/readers/configurations/GdprConfigurationReader.swift
// Copyright (c) 2022-present Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

import Foundation
import SnowplowTracker

struct GdprConfigurationReader: Decodable {
    let basisForProcessing: String
    let documentId: String
    let documentVersion: String
    let documentDescription: String
    
    var basisForProcessingType: GDPRProcessingBasis {
        switch (basisForProcessing) {
        case "contract":
            return .contract
        case "legal_obligation":
            return .legalObligation
        case "legitimate_interests":
            return .legitimateInterests
        case "public_task":
            return .publicTask
        case "vital_interests":
            return .vitalInterest
        default:
            return .consent
        }
    }
}

extension GdprConfigurationReader {
    func toConfiguration() -> GDPRConfiguration {
        return GDPRConfiguration(
            basis: basisForProcessingType,
            documentId: documentId,
            documentVersion: documentVersion,
            documentDescription: documentDescription
        )
    }
}
